tool/gravity/cli/ops.go

```diff
@@ -87,110 +88,155 @@ func appPackage(env *localenv.LocalEnvironment) error {
 }
 
 func uploadUpdate(env *localenv.LocalEnvironment, opsURL string) error {
-	// create local environment with gravity state dir because the environment
-	// provided above has upgrade tarball as a state dir
-	localStateDir, err := localenv.LocalGravityDir()
+	clusterOperator, err := env.SiteOperator()
 	if err != nil {
-		return trace.Wrap(err)
+		return trace.Wrap(err, "unable to access cluster.\n"+
+			"Use 'gravity status' to check the cluster state and make sure "+
+			"that the cluster DNS is working properly.")
 	}
 
-	defaultEnv, err := localenv.New(localStateDir)
+	cluster, err := clusterOperator.GetLocalSite()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	clusterOperator, err := defaultEnv.SiteOperator()
+	tarballEnv, err := localenv.NewTarballEnvironment(localenv.TarballEnvironmentArgs{
+		License: cluster.License.Raw,
+	})
 	if err != nil {
-		return trace.Wrap(err, "unable to access cluster.\n"+
-			"Use 'gravity status' to check the cluster state and make sure "+
-			"that the cluster DNS is working properly.")
+		return trace.Wrap(err)
 	}
+	defer tarballEnv.Close()
 
-	cluster, err := clusterOperator.GetLocalSite()
+	appPackage, err := install.GetAppPackage(tarballEnv.Apps)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	if cluster.State == ops.SiteStateDegraded {
-		return trace.BadParameter("The cluster is in degraded state so " +
-			"uploading new applications is prohibited. Please check " +
-			"gravity status output and correct the situation before " +
-			"attempting again.")
+	err = canUpload(*cluster, *appPackage)
+	if err != nil {
+		return trace.Wrap(err)
 	}
 
-	var tarballPackages pack.PackageService = env.Packages
-	if cluster.License != nil {
-		parsed, err := license.ParseLicense(cluster.License.Raw)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-
-		encryptionKey := parsed.GetPayload().EncryptionKey
-		if len(encryptionKey) != 0 {
-			tarballPackages = encryptedpack.New(tarballPackages, string(encryptionKey))
-		}
+	clusterPackages, err := env.ClusterPackages()
+	if err != nil {
+		return trace.Wrap(err)
 	}
 
-	clusterPackages, err := defaultEnv.ClusterPackages()
+	clusterApps, err := env.SiteApps()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	clusterApps, err := defaultEnv.SiteApps()
+	installedRuntime := cluster.App.Manifest.Base()
+	if installedRuntime == nil {
+		return trace.BadParameter("failed to determine version of base image")
+	}
+	installedRuntimeVersion, err := installedRuntime.SemVer()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	tarballApps, err := env.AppServiceLocal(localenv.AppConfig{
-		Packages: tarballPackages,
-	})
+	deps, err := getUploadDependencies(tarballEnv, *appPackage, *installedRuntimeVersion)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	appPackage, err := install.GetAppPackage(tarballApps)
+	env.PrintStep("Importing application %v v%v", appPackage.Name, appPackage.Version)
```

Comment:
"Importing cluster image..." Looks like we've missed this one when renaming applications to images.